azure/services/dns/arecords.go

```diff
@@ -16,160 +18,221 @@ import (
 	"github.com/go-logr/logr"
 	capzpublicips "sigs.k8s.io/cluster-api-provider-azure/azure/services/publicips"
 	"sigs.k8s.io/controller-runtime/pkg/log"
+)
 
-	"golang.org/x/exp/slices"
-
-	"github.com/giantswarm/dns-operator-azure/v2/azure"
+const (
+	bastionRecordName   = "bastion"
+	bastion1RecordName  = "bastion1"
+	apiRecordName       = "api"
+	apiserverRecordName = "apiserver"
 )
 
-func (s *Service) calculateMissingARecords(ctx context.Context, logger logr.Logger, currentRecordSets []*armdns.RecordSet) []azure.ARecordSetSpec {
-	desiredRecordSet := s.getDesiredARecords()
-
-	var aRecordsToCreate []azure.ARecordSetSpec
-
-	for _, desiredARecordSet := range desiredRecordSet {
-		recordSetCreationNeeded := true
-		for _, recordSet := range currentRecordSets {
-			if recordSet.Type != nil && *recordSet.Type == RecordSetTypeA &&
-				recordSet.Name != nil && *recordSet.Name == desiredARecordSet.Hostname {
-				logger.Info(
-					fmt.Sprintf("DNS A record '%s' found", desiredARecordSet.Hostname),
-					"DNSZone", s.scope.ClusterDomain(),
-					"hostname", desiredARecordSet.Hostname,
-				)
-				recordSetCreationNeeded = false
+func (s *Service) calculateMissingARecords(ctx context.Context, logger logr.Logger, currentRecordSets []*armdns.RecordSet) ([]*armdns.RecordSet, error) {
+	desiredRecordSet, err := s.getDesiredARecords(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var recordsToCreate []*armdns.RecordSet
+
+	for _, desiredRecordSet := range desiredRecordSet {
+
+		logger.V(1).Info(fmt.Sprintf("compare entries individually - %s", *desiredRecordSet.Name))
+
+		currentRecordSetIndex := slices.IndexFunc(currentRecordSets, func(recordSet *armdns.RecordSet) bool { return *recordSet.Name == *desiredRecordSet.Name })
+		if currentRecordSetIndex == -1 {
+			recordsToCreate = append(recordsToCreate, desiredRecordSet)
+		} else {
+			// remove ProvisioningState from currentRecordSet to make further comparison easier
+			currentRecordSets[currentRecordSetIndex].Properties.ProvisioningState = nil
+
+			// compare ARecords[].IPv4Address
```

Comment:
Do we need to make those kind of comparison on the operator side or can we just rely on Azure handling it ? 

I mean , what if the operator on every reconcile just `update the record` to the spec it is supposed to have and azure can decide if it was changed or not, would that work ?

Reply:
doing it that way, i have the fear that we got throttled very soon (especially when we have more than a few workload clusters).